adaptive/learner/balancing_learner.py

```diff
@@ -113,50 +113,67 @@ def strategy(self, strategy):
                 ' strategy="npoints" is implemented.')
 
     def _ask_and_tell_based_on_loss_improvements(self, n):
-        points = []
-        loss_improvements = []
+        selected = [] # tuples ((learner_index, point), loss_improvement)
+        npoints = [l.npoints + len(l.pending_points) for l in self.learners]
         for _ in range(n):
             improvements_per_learner = []
-            pairs = []
+            points_per_learner = []
```

Comment:
These aren't per learner (I'd only call something "per learner" if it's a list of length equal to the number of learners).

Reply:
Are these `requested`?

Reply:
I've changed it. What do you think?

Reply:
Much better now. I think a reader can now understand what's happening without looking elsewhere.